### ISSUE TITLE
negative rewards (fines for killing passive mobs) affects the achiviements

### ISSUE THREAD

**Issue by meswrex** [Rocologo/MobHunting#94]
It appears that having negative numbers for some rewards means that the hunter bonuses will still be applied. Eg. get fined for each mob, but get the hunter bonus when enough have been killed. If confirmed, can the plugin be updated to turn off hunter bonuses when the number is less than or equal to zero?

**Comment by Rocologo** [comment]
Please  help me testing V3.4.7-SNAPSHOT build 116 or newer.
http://jenkins.lindegaard.one:8180/job/MobHunting/116/

Let me know if it works?

**Comment by meswrex** [comment]
Doesn't seem to have changed; hopefully it's not something with my config. Here are the relevant sections:

```
passive:
  chicken: '-1'
  chicken-cmd: execute @p ~ ~ ~ particle reddust ~ ~1 ~ 0.25 0.25 0.25 1 50|execute @p ~ ~ ~ playsound minecraft:block.anvil.land voice @p
  chicken-cmd-desc: Passive Mob Kill! You have been fined!
  chicken-cmd-run-frequency: 100
  chicken-cmd-run-frequency-base: 100
```
```
achievement_levels:
  chicken_level1: 0
```
```
achievements:
  hunt-begins: 500.0
  hunter1: 1000.0
```

I deleted my mobhunting.db, used \summon Chicken, killed the chicken, then got this message:
```
Special Kill Awarded! MobHunting
Make your first hunt. There are more achievements to get all to do with interesting or unique ways to kill mobs.
You have been awarded 500.00 Emeralds
Special Kill Awarded! MobHunting
Hunt 0 Chickens
You have been awarded 1000.00 Emeralds
[Server: Given [Gold Ingots]*5 to Meswrex]
Enjoy your 5 Gold Ingots
Passive Mob Kill! You have been fined!
```

**Comment by Rocologo** [comment]
You should not change chickel_level1 to 0, set it to 100 again and try again. 

I have not checked what happens if chicken_level1 is 0, so this could easily be the problem now. This will be next step if the above does not work.

**Comment by Rocologo** [comment]
Wait - I will do more testing ...

**Comment by Rocologo** [comment]
I think I have found the bug. Please test Build 117
http://jenkins.lindegaard.one:8180/job/MobHunting/117/

**Comment by meswrex** [comment]
In summary I think it's fixed the hunter level rewards, but I think it is still counted as a hunt from a fresh start (deleted db); hence the initial Hunt Begins reward is given.

If I set `chicken_level1: 100` (default) then I get:
```
Special Kill Awarded! MobHunting
Make your first hunt. There are more achievements to get all to do with interesting or unique ways to kill mobs.
You have been awarded 500.00 Emeralds
Special Kill Progress! MobHunting
1/100
Passive Mob Kill! You have been fined!
```
Using mh_achievements I have `The Hunt Begins` and `Beginner Chicken Hunter`.

If I set `chicken_leve1: 0` (incidentally, as recommended by the instructions in the config: "Level Achievements can be disabled by setting the number to 0"), then I get:

```
Special Kill Awarded! MobHunting
Make your first hunt. There are more achievements to get all to do with interesting or unique ways to kill mobs.
You have been awarded 500.00 Emeralds
Passive Mob Kill! You have been fined!
```

Using mh_achievements I have `The Hunt Begins`.

So should I set `chicken_leve1: 0` or not? It looks like it works when it's zero.

**Comment by Rocologo** [comment]
Sorry for the delay. I have spend some more time on this and done some testing myself. I really think it works now. 

If you set chicken_level1: 0 the player will not get Special Hunter Chicken Achievements, and you will not see this reward in /mh achievements. 
 
Please test build 118 or newer: 
http://jenkins.lindegaard.one:8180/job/MobHunting/

**Comment by meswrex** [comment]
I still get `The Hunt Begins` reward for killing a mob unfortunately. It would be good if negative numbers would not be recognised as a reward, even if starting 'The Hunt'.

The Special Hunter Chicken Achievement is working - with chicken_level1: 0 I no longer get this message. 

No problem for the delay - if anything I really appreciate how well you are supporting this. I see some complain about this and that and don't realise they are getting free software with support! The internet is a strange place where things are sometimes back-to-front of common sense and courtesy. Thanks so much again!

**Comment by Rocologo** [comment]
Thats because i didnt change anything in and all the other achievements....  :-) 

I didnt get that you wanted to cancel all achievements if the mob reward is negative, but i see your point. I take a look at the achievements again.

Im very happy about your appreciation. ☺

**Comment by meswrex** [comment]
Ah ok. I think I see now there are two separate issues: (1) turning off individual achievement tallies where the count reward is negative (now resolved!) and (2) having such also not count in any other reward systems such as the general "the  Hunt" achievement. I can put this latter aspect in as a new ticket if you prefer so that we can close this current one(?) "The Hunt" achievement will really only trigger the once per player as soon as they commence any hunt tally, but I suppose it would be cleaner for this to only trigger on mobs that are to be hunted and tallied. I keep seeing it because I am deleting the database on each test, but once I leave the database alone, our players will never see the reward a second time if they choose to kill fine-able mobs as the first mob kill.

**Comment by Rocologo** [comment]
Yes exactly. There two different achievements in the plugin. One time achievements and progress achievements. This was "invented" by the original developer. You could argue that some of the One time achievements could have been made as progress achievements, but I have spend time on that. 

No need for two ticket. I just need a place where we can communicate. The forums arent good because issues then being mixed with other issues and questions and so. 

I will get back to you in short time

**Comment by Rocologo** [comment]
Hi again.

I have now added a test (prize<0) for all "One time achievements" too. I also found a bug, when the player didn't get a message about the Penalty under some circumstances.

Please test build 119 or newer:
http://jenkins.lindegaard.one:8180/job/MobHunting/119/

**Comment by meswrex** [comment]
Hmm. I still get the `special Kill Awarded! Mobhunting` message and the associated reward. I also get fined as expected for killing a chicken. This is the very first hunt since I am still deleting the database prior to each test.

**Comment by Rocologo** [comment]
Hmm. I have done the same testing as you. I delete the sqlite database before starting my test server and it works for me. 

It must be differences in our config.yml, This is mine: 
http://pastebin.com/8Vw5wPpW
Can you see any differences?

I will test again....

**Comment by Rocologo** [comment]
Please enable debug information (/mh debug)

When I kill first mob (a chicken) I get this in the log.

> [16:43:18 INFO]: UUID of player Rocologo is 3c9a8292-0fa7-4e7c-8413-2bcb15968204
> [16:43:21 INFO]: [MobHunting][Debug] Reading Playersettings from Database: PlayerStore: {player: Id:1 Name:Rocologo, Learning Mode: false, Mute Mode: false}
> [16:43:21 INFO]: Rocologo[/127.0.0.1:51162] logged in with entity id 746 at ([world]-134.88655259633995, 63.0, -26.933217205488294)
> [16:43:21 INFO]: [MobHunting][Debug] 0 bounties for Rocologo was loaded.
> [16:43:23 INFO]: [MobHunting][Debug] You are using a development version (3.4.7-SNAPSHOT)
> [16:43:34 INFO]: Rocologo issued server command: /we cui
> [16:44:10 INFO]: Rocologo issued server command: /spawnmob chicken
> [16:44:12 INFO]: [PT] Rocologo issued server command: /fireball
> [16:44:13 INFO]: [MobHunting][Debug] KillAllowed:(1) Rocologo is hiding in WG region, but this is overruled with MobHunting=allow
> [16:44:13 INFO]: [MobHunting][Debug] Permission 'mobhunting.mobs.*' or 'mobhunting.mobs.CHICKEN' not set, defaulting to True.
> [16:44:13 INFO]: [MobHunting][Debug] lastKillstreakLevel=0
> [16:44:13 INFO]: [MobHunting][Debug] newKillstreakLevel=0
> [16:44:13 INFO]: [MobHunting][Debug] Checking if player is grinding mob in the same region within a range of 15.0 blocks
> [16:44:13 INFO]: [MobHunting][Debug] DampendKills=0
> [16:44:13 INFO]: [MobHunting][Debug] Mob Basic Prize=-1.0 for killing a Chicken
> [16:44:13 INFO]: [MobHunting][Debug] Multiplier: §7Rank = 1.05
> [16:44:13 INFO]: [MobHunting][Debug] DifficultyMultiplier: difficulty.multiplier.easy=0.75
> [16:44:13 INFO]: [MobHunting][Debug] WorldDifficulty=EASY, multiplier=0.75
> [16:44:13 INFO]: [MobHunting][Debug] Multiplier: Difficulty = 0.75
> [16:44:13 INFO]: [MobHunting][Debug] Killstreak=1, level=0, multiplier=1.0 
> [16:44:13 INFO]: [MobHunting][Debug] [AchievementBlocked] ProgressAchievement for killing a Chicken Head is disabled (Chicken Head_level1) is 0 in config.yml
> [16:44:13 INFO]: [MobHunting][Debug] Rocologo got a penalty ($-0.79)
> [16:44:13 INFO]: [MobHunting][Debug] RecordKill: Rocologo killed a Chicken
> [16:44:13 INFO]: [MobHunting][Debug] random double=0.8733103909980101 < chance=0.0

What do you get?

**Comment by Rocologo** [comment]
And maybe I did something wrong when I updated my Jenkins server. Please try build 120.

http://jenkins.lindegaard.one:8180/job/MobHunting/120/

**Comment by meswrex** [comment]
Here is the debug output:
```

[07:38:33] [Server thread/INFO]: [MobHunting] Loading MobHunting v3.4.7-SNAPSHOT

...

[07:38:59] [Server thread/INFO]: [MobHunting] Enabling MobHunting v3.4.7-SNAPSHOT
[07:38:59] [Server thread/INFO]: [MobHunting] Enabling compatibility with Citizens (2.0.19-SNAPSHOT (build 1374))
[07:38:59] [Server thread/INFO]: [MobHunting] Enabling Compatibility with Gringotts (2.9)

...

[07:39:00] [Server thread/INFO]: [MobHunting] Checking for new updates...

...

[07:39:06] [Server thread/INFO]: [MobHunting] No update found.
[07:39:09] [Server thread/WARN]: [AutoSaveWorld] Plugin MobHunting attempted to establish connection http://mcstats.org/ in main server thread
[07:39:09] [Server thread/WARN]: [AutoSaveWorld] Plugin MobHunting attempted to establish connection socket://mcstats.org:80 in main server thread

...

[07:44:06] [Server thread/INFO]: Meswrex issued server command: /mh debug
[07:44:21] [Server thread/INFO]: Meswrex issued server command: /summon Chicken
[07:44:21] [Server thread/INFO]: [Meswrex: Object successfully summoned]
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Permission 'mobhunting.mobs.*' or 'mobhunting.mobs.CHICKEN' not
set, defaulting to True.
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] lastKillstreakLevel=0
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] newKillstreakLevel=0
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Mob Basic Prize=-1.0 for killing a Chicken
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] DifficultyMultiplier: difficulty.multiplier.easy=1
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] WorldDifficulty=EASY, multiplier=1
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Multiplier: ?9Sneaky! = 2.0
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Meswrex has no Rank Multiplier
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Killstreak=1, level=0, multiplier=1.0
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] [AchievementBlocked] ProgressAchievement for killing a Chicken Head is disabled (Chicken Head_level1) is 0 in config.yml
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Meswrex got a penalty (-2.00 Emeralds)
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] RecordKill: Meswrex killed a Chicken
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] random double=0.07846411317272095 < chance=1.0
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] command to be run is:execute @p ~ ~ ~ particle reddust ~ ~1 ~ 0.25 0.25 0.25 1 50|execute @p ~ ~ ~ playsound minecraft:block.anvil.land voice @p|mobhunt head give Meswrex Chicken
[07:44:25] [Server thread/INFO]: [Meswrex: Playing effect reddust for 50 times]
[07:44:25] [Server thread/INFO]: Playing effect reddust for 50 times
[07:44:25] [Server thread/INFO]: [Meswrex: Played sound 'minecraft:block.anvil.land' to Meswrex]
[07:44:25] [Server thread/INFO]: Played sound 'minecraft:block.anvil.land' to Meswrex
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Chicken Head Cmd=minecraft:give Meswrex skull 1 3 {SkullOwner:"MHF_Chicken",display:{Name:"Chicken Head",Lore:[MobHunting Reward]}}
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] You dropped a MH Head DisplayName=Chicken Head
[07:44:25] [Server thread/INFO]: Given [Chicken Head] * 1 to Meswrex
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] Description to be send:?a?oPassive Mob Kill! You have been fined 1 emerald!
[07:44:25] [Server thread/INFO]: [MobHunting][Debug] You picked up a MH Head DisplayName=Chicken Head
```
Here is the chat output:
```
Object successfully summoned
Special Kill Awarded! MobHunting
Make your first hunt. There are more achievements to
get all to do with interesting or unique ways to kill mobs.
You have been awarded 64.00 Emeralds
[Server: Given [Chicken Head]*1 to Meswrex]
Passive Mob Kill! You have been fined 1 emerald!
```

**Comment by Rocologo** [comment]
Perfect, i can see whats wrong now. You are not using build 120. Check that you dont have two jars in the plugin folder.

**Comment by meswrex** [comment]
Great work! That was it, thanks. When I tried to delete the prior version I accidentally activated the rename mode before I could stop hitting the delete and enter key quickly. This renamed it as .jar which then became invisible without a ls -a. I couldn't see it so thought it was gone.

All working now. Curiously I get a message `You lost -2.00 Emeralds! With x2.0 * Sneaky!` but I think this is a feature I have to read up on to understand why I got this double penalty. But no more tally or one-off reward.

Thanks so much Rocologo! Wonderful development and support!!!

**Comment by Rocologo** [comment]
:-) it is not first time I have been struggling with a bug, not understanding why my code didnt work, and then later discover that I had 2xjar in my plugins folder.... 

You lost -2 because the sneaky multiplier was activated. I cant remember exactly when sneaky is activated. All mutipliers are multiplied with the prize - even when the prize is negative. I think this is most fair. On my test server the penalty became smaler than -1 (-0.75) Because world difficulty is easy and multiplier for this is 0.75 (default)

**Comment by meswrex** [comment]
Wonderful. Thanks so much again!
```
The Programmer's Blessing:
May your code be small as the acorn, strong as the oak, simple as the wheel, and swift as the wind.
```

**Comment by Rocologo** [comment]
Case closed ☺
